### PR TITLE
feat: Additional EventMessage Handlers

### LINKED ISSUE #25: Create additional EventMessage Handlers
The required eventhandlers were added with #18.
This issue should cover all eventhandlers, which are not required in the short term

# Should contain
- [x] GainExperience (for medic XP etc)
- [x] PlayerFacilityCapture
- [x] PlayerFacilityDefend
- [x] ContinentUnlock (currently a dead event in the API)
- [x] AchievementEarned
- [x] BattleRankUp

### PR DESCRIPTION
**Pull request by marci4**
Fixes #25 

- [x] GainExperience (for medic XP etc)
- [X] PlayerFacilityCapture
- [x] PlayerFacilityDefend
- [x] ContinentUnlock (currently a dead event in the API)
- [x] AchievementEarned
- [x] BattleRankUp

### CODE REVIEW

**Comment by marci4** [comment]
@Maelstromeous I fu**ed up the first push -.- sorry for that

**Comment by marci4** [comment]
I think we can close #18 after this.

File: app/src/handlers/census/events/AchievementEarnedEvent.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+@injectable()
+export default class AchievementEarnedEvent {
+    public readonly worldId: number;
+    public readonly characterId: number;
+    public readonly zone: Zone;
+    public readonly achievementId: number;
+    public readonly timestamp: number;
+
+    public constructor(
+        event: GenericEvent
```
Comment (Maelstromeous) [inline app/src/handlers/census/events/AchievementEarnedEvent.ts:18]: Just thought actually, we should probably change generic event to `CensusEvent`, better naming for it.

File: app/src/handlers/census/events/AchievementEarnedEvent.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        if (isNaN(this.characterId)) {
+            throw new IllegalArgumentException('character_id');
+        }
+        this.achievementId = Parser.parseArgumentAsNumber(achievementEarnedEvent.achievement_id);
+        if (isNaN(this.achievementId)) {
+            throw new IllegalArgumentException('achievement_id');
+        }
+        this.timestamp = Parser.parseArgumentAsNumber(achievementEarnedEvent.timestamp);
+        if (isNaN(this.timestamp)) {
+            throw new IllegalArgumentException('timestamp');
```
Comment (Maelstromeous) [inline app/src/handlers/census/events/AchievementEarnedEvent.ts:37]: Needs origins

File: app/src/handlers/census/events/BattleRankUpEvent.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        if (isNaN(this.characterId)) {
+            throw new IllegalArgumentException('character_id');
+        }
+        this.timestamp = Parser.parseArgumentAsNumber(battleRankUpEvent.timestamp);
+        if (isNaN(this.timestamp)) {
+            throw new IllegalArgumentException('timestamp');
+        }
+        this.battleRank = Parser.parseArgumentAsNumber(battleRankUpEvent.battle_rank);
+        if (isNaN(this.battleRank)) {
+            throw new IllegalArgumentException('battle_rank');
```
Comment (Maelstromeous) [inline app/src/handlers/census/events/BattleRankUpEvent.ts:37]: Needs origins

File: app/src/handlers/census/events/ContinentUnlockEvent.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        this.ncPopulation = Parser.parseArgumentAsNumber(continentUnlockEvent.nc_population);
+        if (isNaN(this.ncPopulation)) {
+            throw new IllegalArgumentException('nc_population');
+        }
+        this.vsPopulation = Parser.parseArgumentAsNumber(continentUnlockEvent.vs_population);
+        if (isNaN(this.vsPopulation)) {
+            throw new IllegalArgumentException('vs_population');
+        }
+        this.trPopulation = Parser.parseArgumentAsNumber(continentUnlockEvent.tr_population);
+        if (isNaN(this.trPopulation)) {
```
Comment (Maelstromeous) [inline app/src/handlers/census/events/ContinentUnlockEvent.ts:49]: Needs origins (check the other event classes too, might have forgot to pick up on this last review)

File: app/src/handlers/census/AchievementEarnedHandler.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/**
```
Comment (Maelstromeous) [inline app/src/handlers/census/AchievementEarnedHandler.ts:1]: Can you move the comments detailing the census events structure out of the handlers and into the event classes themselves, as we don't fully need to know the structure of the census event at the handler level as it's actually parses at the event level.

File: app/src/interfaces/PlayerHandlerInterface.ts [hunk 6]
```diff
@@ -1,11 +1,10 @@
...
 import PlayerLogoutEvent from '../handlers/census/events/PlayerLogoutEvent';
-import GainExperienceEvent from '../handlers/census/events/GainExperienceEvent';
 
 export default interface PlayerHandlerInterface {
     handleLogin(loginEvent: PlayerLoginEvent): boolean;
 
     handleLogout(logoutEvent: PlayerLogoutEvent): boolean;
 
-    handleExperienceEvent(gainExperienceEvent: GainExperienceEvent): void;
+    updateLastSeen(wordId: number, characterId: number): void;
```
Comment (Maelstromeous) [inline app/src/interfaces/PlayerHandlerInterface.ts:9]: This is needed on death event too

File: app/src/handlers/census/events/AchievementEarnedEvent.ts (hunk 1 above)
Comment (marci4) [reply app/src/handlers/census/events/AchievementEarnedEvent.ts:18]: PS2Census may want to change it then

Comment (Maelstromeous) [reply app/src/handlers/census/events/AchievementEarnedEvent.ts:18]: Oh yeah that's coming from there. I'll mention it to microwave.

File: app/src/handlers/census/AchievementEarnedHandler.ts (hunk 5 above)
Comment (marci4) [reply app/src/handlers/census/AchievementEarnedHandler.ts:1]: Done

File: app/src/interfaces/PlayerHandlerInterface.ts (hunk 6 above)
Comment (marci4) [reply app/src/interfaces/PlayerHandlerInterface.ts:9]: Done, implemented DeathEvent

**Comment by Maelstromeous** [comment]
@marci4 conflicts are found, once you fix though we're good :)

File: app/src/handlers/census/events/GainExperienceEvent.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import { injectable } from 'inversify';
```
Comment (Maelstromeous) [inline app/src/handlers/census/events/GainExperienceEvent.ts:1]: Seems to be missing the event details? Is it in the handler?

File: app/src/handlers/census/GainExperienceHandler.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
+/**
```
Comment (Maelstromeous) [inline app/src/handlers/census/GainExperienceHandler.ts:1]: Needs moving to event

**Pull request merged**